### PR TITLE
BrowseModeDocumentTextInfo.getControlFieldSpeech: Report the name of groupings (such as fieldsets) for quicknav and focus jumps, similar to how landmark labels are reported.

### LINKED ISSUE #3321: Fieldsets that contain a checkbox grouping does not read the legend at times
**Reported by nshah9 on 2013-07-01 19:41**
I have 2 pages with checkbox groupings.

On one of the pages, when the grouping receives focus, the screen reader reads the legend as expected. On this page, the checkbox grouping is surrounded by a series of other inputs.

However, on the second page, when the grouping receives focus, the screen reader does not read the legend, which I think is an NVDA bug. If the grouping receives focus from the bottom, i.e (Shift+Tab), the screen reader reads the legend correctly. On this page, the checkbox grouping has a link before it and other inputs after it.

### LINKED ISSUE #6979: NVDA fails to read aria-labeledby label associated with a radio group in browse mode
When using "f" or "r" keys to find radio buttons on a page while in browse mode, NVDA does not read the associated label. At this time I can't say whether this is limited to labels associated by aria-labeledby and other techniques. My current testing example is only with aria-labeledby. My recollection is that all labels are not read. This occurs in all browsers. Tested with IE 11, FF 52, Edge on insider build 15058 and Chrome Canary.

* Actual behavior: the label is only read with focus mode enabled.
* Expected behavior: label should be read the first time the user encounters the radio group similar to what occurs in focus mode.

STR:
1. Open FF.
2. press alt+d to go to the addressbar.
3. Visit the following link: https://ibma.github.io/Va11yS/ARIA/ARIA2_example3.html
4. Navigate to the radio group by repeatedly pressing "f" or pressing "r."

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
Fixes #3321
Fixes #6979 

### Summary of the issue:
When jumping by quicknav or focus in browseMode, legends on fieldsets are not reported.

### Description of how this pull request fixes the issue:
Announce the name of the fieldset (grouping) similar to how we announce the label of landmarks, but only do this for reason_focus so that we do not double speak the name due to it also being reachable by the caret within the content.

### Known issues with pull request:
This will happen fr all groupings with a name in browseMode. Would there be some groupings that have badly calculated names that should not be spoken? Should we limit this specifically to fieldset somehow?
This works in Firefox and Chrome. But not in Edge as they do not correctly set the name of their fieldset groupings. A bug has been filed. This does not work in IE as we would have to add code to calculate the name of a fieldset from its legend ourselves. Is this worth doing?

### Change log entry:
Section: changes
In Browse mode for Firefox and Chrome, the name of form field groups are now announced when moving into them with quick navigation or when tabbing.

### CODE REVIEW

**Review by jcsteh — CHANGES_REQUESTED** [review]
1. I think the "grouping" role should be reported. This makes it more consistent with reporting in focus mode.
2. Consider moving this into `speech.getControlFieldSpeech`. You could put this near the code to support tables (around line 1160).

**Comment by michaelDCurran** [comment]
@jcsteh: ready for another review. But: I noticed this PR introduces double speaking if you are in focus mode due to a field needing passThrough, then tabbing directly into a grouping, where the focus lands on a field that does not require passthrough. E.g. Starting on an edit field, and then tabbing into a fieldset and landing on a checkbox. The grouping is announced twice. This is because focusEntered is processed before switching back to browseMode for the focus. Do we care about this? We'd have to stop focusEntereds when switching to browseMode, but we may have originally done this for a particular reason.

**Review by jcsteh — CHANGES_REQUESTED** [review]
While double speaking is better than what we have now (not speaking potentially important info), i can see this becoming a common gripe, so I think we should fix it now.

> We'd have to stop focusEntereds when switching to browseMode, but we may have originally done this for a particular reason.

I don't think there was any reason other than there being no code to prevent it. The issue is that we get the focusEntered events before the mode switch, since the mode switch happens in gainFocus. We have explicit code to replay focus entered events when switching to focus mode because we ignore focusEntered events if passThrough is False. However, we always handle them when passThrough is True. One approach could be to always ignore them and always let gainFocus handle replay for focus mode.

**Comment by michaelDCurran** [comment]
The double speaking is not at all limited to this grouping case. This can be also reproduced with a list. However,  it is even worse as the problem will only happen the first time, due to yet another issue where focusing a control that requires passThrough, the textInfo speech cache for the treeInterceptor is not updated, thus when it hits another control that uses browse mode, the controlFields are compared  with the last browse mode position.
STR:
Rename  
[test.html.txt](https://github.com/nvaccess/nvda/files/1189506/test.html.txt)
 to html and open in Firefox.
Tab through the controls. First the test link, then the edit field, then the website link.
The website link is in a list.
Notice that the list is spoken both as a focus ancestor, and as semantic info in browse mode.
Shift tab back to the edit field and forward to the website link again.
Notice this time that only the list as a focus ancestor is announced, not the semantic info in browse mode.
Now shift tab back past the edit field to the test link. Notice that "out of list" is announced once reaching the test link.
This very last point will be fixed when #2591 is fixed.
My feeling would be to handle all these issues separately from grouping announcement.

**Comment by jcsteh** [comment]
> My feeling would be to handle all these issues separately from grouping announcement.

Fair enough, though it's worth noting that tabbing is more likely (or at least users are more likely to listen to most of the info) in a form, so encountering this for groupings is arguably more problematic than for lists. In other words, I'm suggesting that groupings might perhaps raise the priority of fixing that problem.

**Review by jcsteh — APPROVED** [review]

File: source/browseMode.py [hunk 1]
```diff
@@ -1284,8 +1284,6 @@ def script_shiftTab(self, gesture):
 	def event_focusEntered(self,obj,nextHandler):
 		if obj==self.rootNVDAObject:
 			self._enteringFromOutside = True
-		if self.passThrough:
-			 nextHandler()
```
Comment (jcsteh) [inline source/browseMode.py]: I think it'd be good to have a comment here explaining that we ignore these events here, but event_gainFocus will replay them if appropriate.

**Review by jcsteh — CHANGES_REQUESTED** [review]
Nice. If possible, I'd prefer to see REASON_ONLYCACHE used rather than a separate onlyCache argument for consistency. If this causes major problems, though, I could probably be convinced to let this fly. :)

**Review by jcsteh — APPROVED** [review]

**Pull request merged**